### PR TITLE
Implement GPU accelerated coordinate descent algorithm

### PR DESCRIPTION
**Pull request by RAMitchell**

### CODE REVIEW

**Comment by RAMitchell** [comment]
Some benchmarking results on an i7 4 core CPU vs. Pascal Titan X below.

The gradient calculation and residual update are much faster than on the CPU. The primary bottlenecks by far are now from calling InitColAccess() and prediction on the CPU. I may do some work re-factoring DMatrix before addressing these.

|    | Dataset           | Num features | Num rows | Density                | Type           | L1 penalty | L2 penalty | Solver                      | Metric         | Accuracy            | Time               | Iterations | Zero Coefficients | 
|----|-------------------|--------------|----------|------------------------|----------------|------------|------------|-----------------------------|----------------|---------------------|--------------------|------------|-------------------| 
| 0  | YearPredictionMSD | 90           | 515345   | 1.0                    | Regression     | 0.05       | 0.1        | XGBoost (coord_descent)     | RMSE           | 0.9239125894920246  | 0.8073031902313232 | 8          | 71/90             | 
| 1  | YearPredictionMSD | 90           | 515345   | 1.0                    | Regression     | 0.05       | 0.1        | XGBoost (shotgun)           | RMSE           | 0.9239179075795608  | 0.9293212890625    | 8          | 71/90             | 
| 2  | YearPredictionMSD | 90           | 515345   | 1.0                    | Regression     | 0.05       | 0.1        | XGBoost (gpu_coord_descent) | RMSE           | 0.9239125888786826  | 0.5511066913604736 | 8          | 71/90             | 
| 3  | YearPredictionMSD | 90           | 515345   | 1.0                    | Regression     | 0.05       | 0.1        | Sklearn                     | RMSE           | 0.9240375468186954  | 1.0513970851898193 | 22         | 71/90             | 
| 4  | YearPredictionMSD | 90           | 515345   | 1.0                    | Regression     | 0.05       | 0.1        | Glmnet                      | RMSE           | 0.9240373005277546  | 5.280339479446411  | -          | 71/90             | 
| 5  | Synthetic         | 100          | 10000000 | 1.0                    | Regression     | 0.05       | 0.1        | XGBoost (coord_descent)     | RMSE           | 0.22794622226312158 | 14.853165864944458 | 5          | 91/100            | 
| 6  | Synthetic         | 100          | 10000000 | 1.0                    | Regression     | 0.05       | 0.1        | XGBoost (shotgun)           | RMSE           | 0.22794622136088685 | 17.738893032073975 | 5          | 91/100            | 
| 7  | Synthetic         | 100          | 10000000 | 1.0                    | Regression     | 0.05       | 0.1        | XGBoost (gpu_coord_descent) | RMSE           | 0.2279462204610238  | 11.507866621017456 | 5          | 91/100            | 
| 8  | Synthetic         | 100          | 10000000 | 1.0                    | Regression     | 0.05       | 0.1        | Sklearn                     | RMSE           | 0.22794621563959372 | 35.15162372589111  | 3          | 91/100            | 
| 9  | Synthetic         | 100          | 10000000 | 1.0                    | Regression     | 0.05       | 0.1        | Glmnet                      | RMSE           | 0.22794621563954823 | 96.25374722480774  | -          | 91/100            | 
| 10 | Higgs             | 28           | 10999999 | 1.0                    | Classification | 0.0        | 0.1        | XGBoost (coord_descent)     | Accuracy score | 0.6222706020246002  | 8.977984189987183  | 12         | 0/28              | 
| 11 | Higgs             | 28           | 10999999 | 1.0                    | Classification | 0.0        | 0.1        | XGBoost (shotgun)           | Accuracy score | 0.6222720565701869  | 8.188422441482544  | 13         | 0/28              | 
| 12 | Higgs             | 28           | 10999999 | 1.0                    | Classification | 0.0        | 0.1        | XGBoost (gpu_coord_descent) | Accuracy score | 0.6222706020246002  | 3.8459713459014893 | 12         | 0/28              | 
| 13 | Higgs             | 28           | 10999999 | 1.0                    | Classification | 0.0        | 0.1        | Sklearn (liblinear)         | Accuracy score | 0.6222183292925754  | 68.72119665145874  | [11]       | 0/28              | 
| 14 | Higgs             | 28           | 10999999 | 1.0                    | Classification | 0.0        | 0.1        | Sklearn (sag)               | Accuracy score | 0.6222603292963935  | 67.09813499450684  | [9]        | 0/28              | 
| 15 | Higgs             | 28           | 10999999 | 1.0                    | Classification | 0.0        | 0.1        | Sklearn (lbfgs)             | Accuracy score | 0.6222756929341539  | 13.75407338142395  | [9]        | 0/28              | 
| 16 | Higgs             | 28           | 10999999 | 1.0                    | Classification | 0.0        | 0.1        | Glmnet                      | Accuracy score | 0.6222748747522613  | 72.70484495162964  | -          | 0/28              | 
| 17 | Higgs             | 28           | 10999999 | 1.0                    | Classification | 0.02       | 0.0        | XGBoost (coord_descent)     | Accuracy score | 0.6127225102475009  | 8.401798963546753  | 16         | 20/28             | 
| 18 | Higgs             | 28           | 10999999 | 1.0                    | Classification | 0.02       | 0.0        | XGBoost (shotgun)           | Accuracy score | 0.6127220557020051  | 5.0337748527526855 | 7          | 20/28             | 
| 19 | Higgs             | 28           | 10999999 | 1.0                    | Classification | 0.02       | 0.0        | XGBoost (gpu_coord_descent) | Accuracy score | 0.6127225102475009  | 4.684044599533081  | 16         | 20/28             | 
| 20 | Higgs             | 28           | 10999999 | 1.0                    | Classification | 0.02       | 0.0        | Sklearn (liblinear)         | Accuracy score | 0.6126702375154761  | 20.47611403465271  | [9]        | 20/28             | 
| 21 | Higgs             | 28           | 10999999 | 1.0                    | Classification | 0.02       | 0.0        | Glmnet                      | Accuracy score | 0.6127227829747984  | 48.62993264198303  | -          | 20/28             | 
| 22 | Cover type        | 54           | 581012   | 1.0                    | Classification | 0.0        | 0.1        | XGBoost (coord_descent)     | Accuracy score | 0.682293997370106   | 14.041389465332031 | 7          | 0/439             | 
| 23 | Cover type        | 54           | 581012   | 1.0                    | Classification | 0.0        | 0.1        | XGBoost (shotgun)           | Accuracy score | 0.6822819494261736  | 7.361542224884033  | 7          | 0/439             | 
| 24 | Cover type        | 54           | 581012   | 1.0                    | Classification | 0.0        | 0.1        | XGBoost (gpu_coord_descent) | Accuracy score | 0.682293997370106   | 1.9714512825012207 | 7          | 0/439             | 
| 25 | Cover type        | 54           | 581012   | 1.0                    | Classification | 0.0        | 0.1        | Sklearn (sag)               | Accuracy score | 0.6823163721231231  | 64.82465410232544  | [52]       | 0/378             | 
| 26 | Cover type        | 54           | 581012   | 1.0                    | Classification | 0.0        | 0.1        | Sklearn (lbfgs)             | Accuracy score | 0.6823180932579705  | 10.180290937423706 | [39]       | 0/378             | 
| 27 | Cover type        | 54           | 581012   | 1.0                    | Classification | 0.0        | 0.1        | Glmnet                      | Accuracy score | 0.6822647380776989  | 52.03886103630066  | -          | 0/378             | 
| 28 | URL Reputation    | 3231961      | 2376130  | 3.5777623485370876e-05 | Classification | 0.0        | 1e-06      | XGBoost (coord_descent)     | Accuracy score | 0.9935117186349232  | 32.19034743309021  | 5          | 17596/3231961     | 
| 29 | URL Reputation    | 3231961      | 2376130  | 3.5777623485370876e-05 | Classification | 0.0        | 1e-06      | XGBoost (shotgun)           | Accuracy score | 0.9945171350052396  | 196.71638560295105 | 118        | 17596/3231961     | 
| 30 | URL Reputation    | 3231961      | 2376130  | 3.5777623485370876e-05 | Classification | 0.0        | 1e-06      | XGBoost (gpu_coord_descent) | Accuracy score | 0.9935117186349232  | 488.0377948284149  | 5          | 17597/3231961     | 
| 31 | URL Reputation    | 3231961      | 2376130  | 3.5777623485370876e-05 | Classification | 0.0        | 1e-06      | Sklearn (liblinear)         | Accuracy score | 0.9846473046508398  | 148.72706246376038 | [24]       | 17596/3231961     | 
| 32 | URL Reputation    | 3231961      | 2376130  | 3.5777623485370876e-05 | Classification | 0.0        | 1e-06      | Sklearn (sag)               | Accuracy score | 0.9944859919280511  | 393.1098008155823  | [100]      | 17596/3231961     | 
| 33 | URL Reputation    | 3231961      | 2376130  | 3.5777623485370876e-05 | Classification | 0.0        | 1e-06      | Sklearn (lbfgs)             | Accuracy score | 0.985930062749092   | 149.67726397514343 | [100]      | 17596/3231961     |

**Comment by RAMitchell** [comment]
There are still issues with some of the GPU unit tests and multi-GPU has not been properly tested. 

I will be at a conference over the next two weeks so will resolve these when I get back.

**Comment by khotilov** [comment]
Some observations, primarily regarding the benchmarking:
- Overall, the GPU updater seems to be working quite good! However, there also seems to be some inefficiency in the case with lots of sparse features.
- The regularization parameters were not selected to be close to optimal for the datasets.
- The glmnet looks quite slower b/c in the python implementation there's an internal 3-fold CV for selecting best lambda (which is not really doing anything useful here, since you force lambda_path to a single value).
- Using the early stopping in training data in xgboost doesn't really detect the fit convergence, it only detects that the metric used for early stopping (which is usually not the fit objective) happened to be the best at some iteration (which is more likely to happen with eta=1 and strong regularization), but it might take some more iterations for the fit objective to stabilize. Also, the "Accuracy" column looking better for xgboost is mostly an effect of overfitting with the early stopping.

**Comment by RAMitchell** [comment]
@khotilov 

> Overall, the GPU updater seems to be working quite good! However, there also seems to be some inefficiency in the case with lots of sparse features.

The gpu updater may never be optimal in this case due to the overhead of launching kernels for very small amounts of work. Nvidia has recently dramatically improved kernel launch overhead in cuda 9.1/9.2 so it will be interesting to see the effect here. Perhaps we can just fall back to the cpu in this extreme case.

> The regularization parameters were not selected to be close to optimal for the datasets.

I have simply selected these parameters to expose the effect of sparsity on runtime. Not sure if it matters if they are optimal but this could be done.

>The glmnet looks quite slower b/c in the python implementation there's an internal 3-fold CV for selecting best lambda (which is not really doing anything useful here, since you force lambda_path to a single value).

I will have to find a way around this.

> Using the early stopping in training data in xgboost doesn't really detect the fit convergence, it only detects that the metric used for early stopping (which is usually not the fit objective) happened to be the best at some iteration (which is more likely to happen with eta=1 and strong regularization), but it might take some more iterations for the fit objective to stabilize. Also, the "Accuracy" column looking better for xgboost is mostly an effect of overfitting with the early stopping.

I think I will try a python callback to do early stopping based on the size of the model update in one iteration.

You also mentioned you had some more ideas for GLM - do you want to discuss those at some point, maybe over Google hangouts?

File: src/common/device_helpers.cuh [hunk 1]
```diff
@@ -968,4 +973,53 @@ class AllReducer {
...
+ * \tparam  func_t  Type of the function t.
+ * \param shards  The shards.
+ * \param f       The func_t to process.
+ */
+
+template <typename T, typename func_t>
+void ExecuteShards(std::vector<T> *shards, func_t f) {
+  auto previous_num_threads = omp_get_max_threads();
+  omp_set_num_threads(shards->size());
+#pragma omp parallel
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply src/common/device_helpers.cuh]: It does seem like omp_get_num_threads() can safely be set to larger than the number of physical cores:
https://stackoverflow.com/questions/11691960/omp-more-threads-than-the-number-of-processors

File: src/common/device_helpers.cuh [hunk 2]
```diff
@@ -929,14 +931,17 @@ class AllReducer {
...
    *
-   * \param           communication_group_idx Zero-based index of the communication group. \param
-   *                                          sendbuff                The sendbuff. \param sendbuff
-   *                                          The sendbuff. \param [in,out]  recvbuff The recvbuff.
-   *                                          \param           count                   Number of.
+   * \param           communication_group_idx Zero-based index of the
+   * communication group. \param sendbuff                The sendbuff. \param
+   * sendbuff The sendbuff. \param
+   * [in,out]  recvbuff The recvbuff.
+   *                                          \param           count Number of.
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply src/common/device_helpers.cuh]: Fixed thanks.
